File: Data Structures/LinkedLists/LinkedList.cpp
#include <iostream>
using namespace std;
struct Node {
    int data;
    struct Node *next;
};
class LinkedList {
    public:
    Node *head, *tail;
    LinkedList(){
        head = NULL;
        tail = NULL;
    }
	void add(int data) {
        Node *temp = new Node;
        temp->data = data;
        temp->next = NULL;
        if(head == NULL) {
            head = tail = temp;
        }
        else {
            tail->next = temp;
            tail = temp;
        }
	}
    void add(int data, int index) {
        Node *pre = new Node;
        Node *cur = new Node;
        Node *temp = new Node;
        temp->data = data;
        cur = head;
        for(int i=0;i<index+1;i++){
            pre = cur;
            cur = cur->next;
        }
        pre->next = temp;
        temp->next = cur;
    }
    void addFirst(int data) {
        Node *temp = new Node;
        temp->data = data;
        temp->next = head;
        head = temp;
    }
    void remove(){
        Node *pre = new Node;
        Node *cur = new Node;
        cur = head;
        while(cur->next != NULL){
            pre = cur;
            cur = cur->next;
        }
        pre->next =NULL;
        delete cur;
    }
    void remove(int index) {
        Node *pre = new Node;
        Node *cur = new Node;
        cur = head;
        for (int i = 0; i < index+1; i++) {
            pre = cur;
            cur = cur->next;
        }
        pre->next = cur->next;
        delete cur;
    }
    void removeFirst(){
        Node *first = new Node;
        first = head;
        head = head->next;
        delete first;
    }
    void print() {
        Node *temp = new Node;
        temp = head;
        while(temp!=NULL){
            cout<<temp->data<<" ";
            temp = temp->next;
        }
        cout<<endl;
    }

};
int main(){
    LinkedList list;
    int n;
    cout<<"Input size: ";
    cin>>n;
    while(n--){
        int t;
        cin>>t;
        list.add(t);
    }
    list.print();
    list.add(10);
    list.addFirst(10);
    list.print();
    list.add(6,6);
    list.print();
    list.remove(6);
    list.remove();
    list.removeFirst();
    list.print();
}

File: Data Structures/LinkedLists/linked_list.py
class Node:

    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:

    def __init__(self):
        self.root = None

    def add_first(self, data):
        node = Node(data)
        node.next = self.root
        self.root = node
    
    def add_last(self, data):
        node = Node(data)

        if self.root is None:
            self.root = node
        else:
            current = self.root
            while current.next is not None:
                current = current.next
            current.next = node

    def delete_first(self):
        self.root = self.root.next

    def delete_last(self):
        current = self.root
        mem = None

        if current is None:
            print('Underflow')
        else:
            while current.next is not None:
                mem = current
                current = current.next
            mem.next = None

    def print(self):
        current = self.root
        while current is not None:
            print(current.data, end=' ')
            current = current.next
        print()

    def size(self):
        current = self.root
        size = 0

        while current is not None:
            size += 1
            current = current.next
        
        return size

if __name__ == "__main__":
    
    l = LinkedList()

    l.add_first(1)
    l.add_first(2)
    l.add_first(3)
    l.add_last(4)
    l.add_last(5)

    l.delete_first()

    print(l.size())

    l.print()
    l.delete_last()
    l.print()

    print(l.size())